Article store events

seanwiseman opened the pull request:
Implements requirements from [#31](https://github.com/libero/walking-skeleton/issues/31).

seanwiseman committed:
Update article tests to use /versions in urls

seanwiseman committed:
Switch dashboard to listen on `articles` queue

seanwiseman committed:
Add pika dependancy

seanwiseman committed:
Add articles.events package

seanwiseman committed:
Add message publisher to article views

seanwiseman committed:
Add broker params to docker-compose.yml

seanwiseman commented:
Currently sends messages for the following events:

`article.delete.[started|completed|failed]`
`article.version.[delete|post|put].[started|completed|failed]`

You should see them populate the dashboard.

thewilkybarkid commented on dashboard/config/packages/old_sound_rabbit_mq.yaml:
```diff
@@ -8,6 +8,6 @@ old_sound_rabbit_mq:
+                type: fanout
             queue_options:
-                name: 'dashboard'
+                name: 'articles'
```
Should be the exchange's name changed?

seanwiseman replied:
As we are not producing any messages (from what I can see) from the dashboard at present, I am not sure of the exchange's purpose here 🤔 .  Maybe was just for testing.

You only need the queue name (`articles` in this case) to receive the messages. The `exchange_options` entry can be removed if we are not producing messages from the dashboard.

thewilkybarkid replied:
Should be the exchange that the queue is bound to (this is under the `consumers` config rather than `producers`).

thewilkybarkid replied:
Can see this is actually working... taking a proper look.

thewilkybarkid replied:
I had worked out what was happening, must not have submitted my follow-up. 🤔

This results in an exchange called 'dashboard' and a bound queue called 'articles', the article store _was_ creating an exchange called 'articles' and a bound queue called 'articles'. This works as the queues matched, the 'dashboad' exchange was just superfluous.

Now the article store doesn't create a queue, this probably no longer works. `exchange_options.name` needs to be  'articles' and `queue_options.name` needs to be 'dashboard'. That binds the queue for this app to the exchange provided by the article store.

seanwiseman replied:
Ok, I have made the change and retested and it works fine.

thewilkybarkid commented on article-store/events/conf.py:
```diff
@@ -0,0 +1,21 @@
+RABBITMQ_USER = os.environ.get('RABBITMQ_USER')
+
+BROKER_CREDENTIALS = pika.PlainCredentials(RABBITMQ_USER, RABBITMQ_PASSWORD)
+BROKER_PARAMS = pika.ConnectionParameters(host=RABBITMQ_HOST, credentials=BROKER_CREDENTIALS)
```
Same as https://github.com/libero/walking-skeleton/pull/25#discussion_r189183325

seanwiseman replied:
Done 👍

thewilkybarkid commented on article-store/events/utils.py:
```diff
@@ -0,0 +1,122 @@
+        raise
+
+
+def setup_exchanges_and_queues(func) -> Callable[..., None]:
```
We don't need a queue here? Though guess this is just copied from earlier work?

seanwiseman replied:
Yes, it was brought in from previous work, but also makes sure the queue(s) you are binding the exchange(s) to exists before the actual binding. This may be subject to change if we can guarantee a queues existence but for now, I think we have to do it this way.

thewilkybarkid replied:
But in this case we don't need the `articles` queue?

seanwiseman replied:
No, it only actually needs an exchange to publish.

seanwiseman committed:
Use broker URL instead of params for rabbitmq

seanwiseman commented:
Are we going to want the `article.version.published` event available in this PR?

giorgiosironi commented on article-store/Pipfile:
```diff
@@ -10,12 +10,14 @@ name = "pypi"
 django-debug-toolbar = "~=1.9.1"
 djangorestframework = "~=3.8.2"
-psycopg2-binary = "*"
+"psycopg2-binary" = "*"
```
pipenv strikes back - not a problem

thewilkybarkid replied:
https://github.com/pypa/pipenv/issues/1933

giorgiosironi commented on article-store/articles/tests/test_api/test_article_api.py:
```diff
@@ -227,7 +235,7 @@ def test_can_get_article_content_without_giving_part_name(self,
                                                               article_0065_es_front_xml: str):
 
-        url = f'/articles/{article.id}/latest'
+        url = f'/articles/{article.id}/versions/latest'
```
changes like these should usually be a separate PR?

giorgiosironi replied:
not that the history of the walking skeleton itself should be that clean in the long run, projects are going to be extracted in other repositories

seanwiseman replied:
Yes I agree.

giorgiosironi commented on article-store/events/conf.py:
```diff
@@ -0,0 +1,17 @@
+
+BROKER_PARAMS = pika.connection.URLParameters(RABBITMQ_URL)
+
+ARTICLES_QUEUE = os.environ.get('ARTICLES_QUEUE', 'articles')
```
I agree with other comments that a queue should not be necessary here. If we want to set up one for testing, it should be clearly named as `articles-dummy` or similar

seanwiseman committed:
Fix: Add default RABBITMQ_URL string value if not present

Without this the article-store-postgres service throws an error because it is not being provided the broker details

giorgiosironi commented on article-store/events/utils.py:
```diff
@@ -0,0 +1,122 @@
+                                 data={'article_version': article_version})
+
+        channel.basic_publish(exchange=ARTICLES_EXCHANGE,
+                              routing_key="",
```
didn't understand this can be empty, good to know

seanwiseman committed:
Renamed ARTICLES_QUEUE to ARTICLES_DUMMY_QUEUE

To still allow for testing

giorgiosironi commented on article-store/events/utils.py:
```diff
@@ -0,0 +1,122 @@
+    return {
+        "eventId": str(uuid.uuid1()),
+        "happenedAt": strftime("%Y-%m-%dT%H:%M:%S+00:00", gmtime()),
+        "aggregate": {
```
may be changed to `entity` in the schema (not yet)

giorgiosironi commented on article-store/events/utils.py:
```diff
@@ -0,0 +1,122 @@
+                channel.queue_declare(queue=queue_name, durable=True)
+
+                for exchange in DEFAULT_QUEUES[queue_name]:
+                    channel.exchange_declare(exchange=exchange, exchange_type='fanout', durable=True)
```
exchanges (only `ARTICLES_EXCHANGE`) should be declared even if there are no queues to be bound to them; this code can be pulled up to the start of the `with get_channel()` block

giorgiosironi commented on article-store/events/utils.py:
```diff
@@ -0,0 +1,122 @@
+    :return:
+    """
+
+    send_message = partial(send_article_message, article_id=article_id, article_version=article_version)
```
this is neat

giorgiosironi replied:
(this context manager in general)

thewilkybarkid replied:
Agreed!

giorgiosironi commented on article-store/events/utils.py:
```diff
@@ -0,0 +1,122 @@
+        "happenedAt": strftime("%Y-%m-%dT%H:%M:%S+00:00", gmtime()),
+        "aggregate": {
+            "service": "article-store",
+            "name": "article-version",
```
may be `article` if this is an `article.delete`-related event?

seanwiseman replied:
mmm, the `article.delete` or `article.version.delete` is currently defined in the `type`, does `aggregate.name` have to change with it as well? or could `aggregate.name` just always be `article`?

giorgiosironi replied:
I'd say comment out the `aggregate` field for now and wait for https://github.com/libero/walking-skeleton/pull/28/files to add it later when stable

seanwiseman replied:
Ok, the article id is in `aggregate.identifier` currently. Should it be moved elsewhere for the time being?

giorgiosironi replied:
No need to do big changes, just adding a comment here about this field being experimental and about to evolve/being replaced would be fine.

seanwiseman committed:
Remove setup of queues, only setup exchanges

thewilkybarkid commented on article-store/events/utils.py:
```diff
@@ -0,0 +1,118 @@
+        yield
+        send_message(f'{msg_type}.completed')
+    except Exception:
+        send_message(f'{msg_type}.failed')
```
Shouldn't this `raise`?

seanwiseman replied:
Good spot.

thewilkybarkid commented on article-store/events/conf.py:
```diff
@@ -3,15 +3,12 @@
-BROKER_PARAMS = pika.connection.URLParameters(RABBITMQ_URL, '')
+BROKER_PARAMS = pika.connection.URLParameters(RABBITMQ_URL)
 
 ARTICLES_DUMMY_QUEUE = os.environ.get('ARTICLES_DUMMY_QUEUE', 'articles-dummy')
```
Can be removed?

seanwiseman committed:
Remove dummy article queue

seanwiseman committed:
Re raise exception in message_publisher

seanwiseman committed:
Configure articles exchange and dashboard queue

giorgiosironi commented on dashboard/config/packages/old_sound_rabbit_mq.yaml:
```diff
@@ -7,7 +7,7 @@ old_sound_rabbit_mq:
-                name: 'dashboard'
-                type: direct
+                name: 'articles'
+                type: fanout
```
not sure why the dashboard should know about the type of the exchange, is it required to subscribe to it? The risk is a race condition where the dashboard preempts the article store in creating the exchange, which shouldn't belong to that service.

thewilkybarkid replied:
Untested, but looks like you can set `declare: false` and it won't create it.

seanwiseman replied:
Trying to test this now. Which level is `declare` configured at in the file?

All combinations I have tried so far have still lead to `You must provide an exchange type  
`.

thewilkybarkid replied:
Same level (https://github.com/php-amqplib/RabbitMqBundle/blob/cf67adaa4e306d8e9cb6855a72d79263b425ded8/DependencyInjection/Configuration.php#L371-L389), so should be:

```yaml
exchange_options:
    name: 'articles'
    type: fanout
    declare: false
```

seanwiseman replied:
Ah ok, thank you. I thought the `declare` meant we did not have to specify a value for the `type`.

OK, this works. It does not create the exchange when `declare: false` is present.

thewilkybarkid replied:
I was surprised too, feels redundant.

seanwiseman replied:
This however causes the following condition: (after a clean startup with no previous volume data)

- `dasboard` queue present
- `articles` exchange not present until the store creates it
- `dashboard` therefore cannot bind to it: (error on startup)
```
In AMQPChannel.php line 188:

NOT_FOUND - no exchange 'articles' in vhost '/'  
```
- There does not seem to be another attempt to bind (though are we expecting it to?)
- Even after a manual binding (via the rabbitMQ UI) of `dashboard` --> `articles` and new events being sent, the dashboard (app) still does not consume the events. I can however see the events pending on the queue (via the rabbitMQ UI) so they are being delivered to the `dashboard` queue.

Will probably benefit from an in person discussion.

giorgiosironi replied:
Yes, we've seen this before https://github.com/libero/walking-skeleton/pull/25/files#diff-bca2c15ea701ffde5b33224920caa539R39

seanwiseman replied:
As discussed, I will not implement the `declare: false`, it will be handled in another pr along with a method of checking the exchange is declared before trying to bind.

seanwiseman commented:
:shipit: ?

giorgiosironi reviewed (approved):
Event data needs to be removed/refined, but let's do it separately not to hold this

thewilkybarkid reviewed (approved):
🎉
